### PR TITLE
Ordering

### PR DESCRIPTION
**Pull request by panvourtsis**
## Description of the PR
This is a solution to the drag and drop ordering.

### Issues Related
This is targeting the #87 issue

------

Please, before doing the PR make sure that you are following the intructions described in [CONTRIBUTING.md](https://github.com/angrykoala/gaucho/blob/master/CONTRIBUTING.md) and you have completed the following checklist

**Checklist**
- [x] Your PR is done against the `dev` branch
- [x] Your origin branch is updated with the latest changes from `dev`
- [x] All the tests and jshint are passing (`npm run test`)
- [ ] Changelog is updated with a brief description of your changes (if required)
- [x] Your changes have tests

Thanks for contributing to Gaucho

### CODE REVIEW

**Comment by angrykoala** [comment]
Hi @panvourtsis Is this the final version of the ordering PR?

**Comment by panvourtsis** [comment]
yes

**Comment by angrykoala** [comment]
I'll do a full review as soon as I can, thanks for your PR

**Comment by angrykoala** [comment]
Ok, after testing it out a bit, seems to be working reasonably good, a few notes on the current behaviour I noted:
* If an edit menu is opened (to edit a task), it should be closed when a reorder starts (ideally it would be automatically be opened afterwards if possible), this is important to allow tasks to be reordered
* A minor glitch where some tasks appear to "jump" sometimes (when there are 3 or more tasks and you try to move one in the middle)

**Comment by angrykoala** [comment]
Please, update the changelog with a new bullet point explaining the changes (a simple one-liner is enough), It doesn't have to be perfectly explained, as I will revise it afterwards, but I think is better that the author describes the feature

File: app/renderer/components/task_card.js [hunk 1]
```diff
@@ -24,9 +24,14 @@ module.exports = {
         "progress-spinner": ProgressSpinner
     },
     template: `
-    <li class="run-card">
+    <li class="run-card task-card">
```
Comment (angrykoala) [inline app/renderer/components/task_card.js]: Why is the class task-card added to the li inside a task-card?

File: app/renderer/components/add_task.js [hunk 2]
```diff
@@ -3,11 +3,13 @@
 const TaskInput = require('./task_input');
 
 module.exports = {
+    props: ['tasks'],
     data() {
         return {
             title: "",
             command: "",
             path: "",
+            order: 0,
```
Comment (angrykoala) [inline app/renderer/components/add_task.js]: Do we really need the order explicitly inside the task? the order should only affect to suites, and the array of tasks already contains that information

File: app/renderer/components/task_input.js [hunk 3]
```diff
@@ -4,12 +4,13 @@ const Task = require('../task');
...
 
 module.exports = {
-    props: ['task'],
+    props: ['task', 'order'],
     data() {
         return {
             title: "",
             command: "",
             path: "",
+            order: 0,
```
Comment (angrykoala) [inline app/renderer/components/task_input.js]: Same as before, I'm not sure we need the explicit order here

File: app/renderer/components/task_suite.js [hunk 4]
```diff
@@ -5,6 +5,7 @@ const EventEmitter = require('events');
 const TaskCard = require('./task_card');
 const AddTask = require('./add_task');
 const AppStatus = require('../app_status');
+const Draggable = require('vuedraggable');
```
Comment (angrykoala) [inline app/renderer/components/task_suite.js]: Please, move the external dependency (vuedraggable) to the top, just for ordering :)

File: app/renderer/components/task_suite.js [hunk 5]
```diff
@@ -39,6 +41,23 @@ module.exports = {
         AppStatus.events.removeListener("stop-suite", this.onStopSuite);
     },
     methods: {
+        checkMove() {
+            if (AppStatus.editMode) {
```
Comment (angrykoala) [inline app/renderer/components/task_suite.js]: Why not `return AppStatus.editMode`?

File: app/renderer/components/task_suite.js [hunk 6]
```diff
@@ -39,6 +41,23 @@ module.exports = {
...
+                return true;
+            }
+            return false;
+        },
+        reStructureTasks(event) {
```
Comment (angrykoala) [inline app/renderer/components/task_suite.js]: restructure is a single word, better naming would be `restructureTasks`

File: app/renderer/components/task_suite.js [hunk 7]
```diff
@@ -71,7 +90,9 @@ module.exports = {
         },
         showAddTab() {
             return AppStatus.editMode && this.suite.length < AppStatus.maxTasksPerSuite;
+        },
+        orderTasks() {
+            return this.suite.tasks.sort((a, b) => a.order - b.order);
```
Comment (angrykoala) [inline app/renderer/components/task_suite.js]: Instead of using the order property of a task, why not just using an array?

File: app/renderer/components/task_card.js (hunk 1 above)
Comment (panvourtsis) [reply app/renderer/components/task_card.js]: this is here because i set on draggable (task_suite.js line 25 -> :options="{draggable:'.task-card'}") that the class of the option will be that to avoid the "add new" li or future options that we dont need

File: app/renderer/components/task_suite.js (hunk 7 above)
Comment (panvourtsis) [reply app/renderer/components/task_suite.js]: not quite sure what do you mean here ...

Comment (angrykoala) [reply app/renderer/components/task_suite.js]: So, atm you have a set of tasks, and within each task you have a explicit property (order), and you use it to well... order.

However, it doesn't make sense to have that property as part of a task (why would a task care of its own position within different tasks?). In fact, a suite has an array/list of tasks, which, by definition, maintains the order (that was done by design at the beginning). So:

Instead of having:
```js
[{
   name: "second task",
   order: 2
}, {
   name: "first task",
   order: 1
}]
```

And ordering them according to its internal value. We could simply take care in maintaining the array order when writing, reading, moving tasks and rendering:
```js
[{
  name: "first task"
}, {
  name: "second task"
}]
```

We do not need to manually store the order and sort according to it. The array data structure already do that for us, doing this way has some advantages:
* No order property in tasks, no manually ordering each time, the array already contains this information, in general, simpler
* Better for storing/reading data, as we do not need to store the ordering property
* It doesn't make sense to store order in the tasks as it is irrelevant for it, moving a task from one place to another shouldn't change the task itself, but its container (suite)

Basically, I have the impression that solving this issue doesn't require changes to the task.js class, as everything should be part of the suite

I'm not sure if I'm explaining myself, sorry, if you have any doubt I'm happy to make a simple prototype/example to show my point

From what I saw in your code it shouldn't be hard to change it to the way I see it, as most changes you made in the suite are well defined, clean and simple enough to adapt them

**Comment by panvourtsis** [comment]
check my latest commit for the changes we discussed. You were right about the ordering, got a bit confused when we had both arrows and drag but now i think its clean. Let me know what you thing

File: app/renderer/task_config.js [hunk 8]
```diff
@@ -74,7 +74,7 @@ module.exports = {
         return data.map((suite) => {
             let result = new Suite(suite.title);
             result.tasks = suite.tasks.map((task) => {
-                return new Task(task.title, task.path, task.command);
+                return new Task(task.title, task.path, task.command, task.order);
```
Comment (angrykoala) [inline app/renderer/task_config.js]: I understand that task.order no longer exists. Is this here for some reason or is a typo?

**Comment by angrykoala** [comment]
Everything apart from that extra parameter looks good in the code @panvourtsis

**Comment by angrykoala** [comment]
The code is good, a few details I noticed when running it:

* As I commented before, when starting a drag **all** the tasks should be closed (the edition menu) as it makes harder to actually reorder:

image

* The "drag" behavior is still active when not on edit mode (you cannot actually change the tasks order, but the cards are draggable, this should be trival to solve, adding a class to avoid this when not in edit mode 
img

* The tasks "jump" when reordering sometimes (sorry, I cannot make a screenshot of this) I'm not sure if this is easy or hard to solve, as I feel it is caused by "removing & adding" a task on drag. It may even be because of the library

I'm happy to create separate issues for these problems, as none of them is critical. But if you could solve the ones that you can, I think it would be faster to develop, as you know and remember all the code now

**Comment by panvourtsis** [comment]
create the tasks and i will work on them, assign them to me if you will. Its easier for me also to have it seperate. Thanks

**Comment by angrykoala** [comment]
Ok, I'll create them but GH doesn't allow me to asisgn them to you, please leave a comment to let people know you are going to work in it
